tools/iotools.go

```diff
@@ -89,3 +93,42 @@ func (r *RetriableReader) Read(b []byte) (int, error) {
 
 	return n, errors.NewRetriableError(err)
 }
+
+// Spool spools the contents from 'from' to 'to' by buffering the entire
+// contents of 'from' into a temprorary file, first.
+//
+// A name of "lfs-<time>" is used, where "<time>" is the number of seconds since
+// the Unix epoch.
+//
+// The temporary file is cleaned up after the copy is complete.
+//
+// The number of bytes written to "to", as well as any error encountered are
+// returned.
+func Spool(to io.Writer, from io.Reader) (n int64, err error) {
+	return SpoolName(to, from, fmt.Sprintf("lfs-%d", time.Now().Unix()))
+}
+
+// Spool spools the contents from 'from' to 'to' by buffering the entire
+// contents of 'from' into a temprorary file named "name", first.
+//
+// The temporary file is cleaned up after the copy is complete.
+//
+// The number of bytes written to "to", as well as any error encountered are
+// returned.
+func SpoolName(to io.Writer, from io.Reader, name string) (n int64, err error) {
+	tmp, err := ioutil.TempFile("", name)
+	if err != nil {
+		return 0, errors.Wrap(err, "spool tmp.")
+	}
+	defer os.Remove(tmp.Name())
+
+	if n, err = io.Copy(tmp, from); err != nil {
```

Comment:
I would imagine that this usually happens to small files because if you add a 1GB file directly to Git instead of LFS then you would realize that pretty quickly.

Therefore I wonder if we should indeed spool the first 10MB in memory to avoid any slow down due to disk I/O?!